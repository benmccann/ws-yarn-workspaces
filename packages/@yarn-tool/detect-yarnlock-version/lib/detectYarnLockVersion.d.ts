/// <reference types="node" />
import { EnumDetectYarnLock, EnumDetectYarnLockInputType } from '@yarn-tool/yarnlock-types';
export declare function _detectYarnLockVersionSimple(buf: Buffer | string): EnumDetectYarnLock;
export declare function _detectYarnLockVersionCore<T extends Buffer | string>(input: T): {
    verType: EnumDetectYarnLock.v1 | EnumDetectYarnLock.v2 | EnumDetectYarnLock.v3;
    detectType: EnumDetectYarnLockInputType.simple;
    input: T;
} | {
    verType: EnumDetectYarnLock.v2 | EnumDetectYarnLock.v3;
    detectType: EnumDetectYarnLockInputType.parse_raw;
    input: T;
} | {
    verType: EnumDetectYarnLock.unknown;
    detectType: EnumDetectYarnLockInputType.unknown;
    input: T;
};
export declare function detectYarnLockVersion(buf: Buffer | string): EnumDetectYarnLock;
/**
 * only check v2 and v3
 */
export declare function _tryParse(buf: Buffer | string): EnumDetectYarnLock.v2 | EnumDetectYarnLock.v3;
export declare function _tryParseObject(yarnLockObject: Record<string, any>): EnumDetectYarnLock.v2 | EnumDetectYarnLock.v3;
export default detectYarnLockVersion;
