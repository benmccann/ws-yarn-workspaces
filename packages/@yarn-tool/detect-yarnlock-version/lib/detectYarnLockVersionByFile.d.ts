import { EnumDetectYarnLock } from './types';
export declare function detectYarnLockVersionByFile(file: string): EnumDetectYarnLock;
export declare function detectYarnLockVersionByDir(dir: string): EnumDetectYarnLock;
export default detectYarnLockVersionByFile;
