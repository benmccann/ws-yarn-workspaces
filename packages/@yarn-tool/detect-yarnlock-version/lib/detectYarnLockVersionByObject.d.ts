import { EnumDetectYarnLock, IYarnLockMetaVersion, IYarnLockVersionSyml } from '@yarn-tool/yarnlock-types';
export declare function detectYarnLockVersionByObject(yarnLockObject: Record<string, any>): EnumDetectYarnLock;
export declare function _checkV2Row(row: unknown): boolean;
/**
 * check v2 and v3
 */
export declare function checkV2(obj: any): EnumDetectYarnLock.unknown | IYarnLockVersionSyml;
export declare function checkV1(obj: any): EnumDetectYarnLock;
export declare function _checkExistsMetaCore(obj: unknown): {
    verType: EnumDetectYarnLock.unknown | IYarnLockVersionSyml;
    metaVersion: `${number}`;
    existsMeta: boolean;
    emptyEntries: boolean;
    existsEntries: boolean;
    onlyExistsMeta: boolean;
    meta: {
        version: IYarnLockMetaVersion;
    };
    entryKeys: string[];
};
export declare function checkExistsMeta(obj: unknown): obj is {
    __metadata: {
        version: IYarnLockMetaVersion;
    };
};
export default detectYarnLockVersionByObject;
