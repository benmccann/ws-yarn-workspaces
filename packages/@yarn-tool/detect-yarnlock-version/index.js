"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectYarnLockVersionByObject = void 0;
__exportStar(require("./lib/detectYarnLockVersion"), exports);
__exportStar(require("./lib/detectYarnLockVersionByFile"), exports);
var detectYarnLockVersionByObject_1 = require("./lib/detectYarnLockVersionByObject");
Object.defineProperty(exports, "detectYarnLockVersionByObject", { enumerable: true, get: function () { return detectYarnLockVersionByObject_1.detectYarnLockVersionByObject; } });
__exportStar(require("./lib/types"), exports);
const detectYarnLockVersion_1 = require("./lib/detectYarnLockVersion");
exports.default = detectYarnLockVersion_1.detectYarnLockVersion;
//# sourceMappingURL=index.js.map