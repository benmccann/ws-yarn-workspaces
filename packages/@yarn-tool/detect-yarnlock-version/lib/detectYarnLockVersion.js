"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectYarnLockVersion = void 0;
const types_1 = require("./types");
function detectYarnLockVersion(buf) {
    let head = buf.slice(0, 160).toString().trim();
    if (head.includes('# yarn lockfile v1')) {
        return types_1.EnumDetectYarnLock.v1;
    }
    else if (/^__metadata:\s*version: 4(?:\r|\n)/m.test(head)) {
        return types_1.EnumDetectYarnLock.berry;
    }
    return types_1.EnumDetectYarnLock.unknown;
}
exports.detectYarnLockVersion = detectYarnLockVersion;
exports.default = detectYarnLockVersion;
//# sourceMappingURL=detectYarnLockVersion.js.map