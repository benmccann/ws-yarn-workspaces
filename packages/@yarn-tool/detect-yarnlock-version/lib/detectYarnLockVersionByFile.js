"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.detectYarnLockVersionByFile = detectYarnLockVersionByFile;
exports.detectYarnLockVersionByDir = detectYarnLockVersionByDir;
const detectYarnLockVersion_1 = require("./detectYarnLockVersion");
const fs_1 = require("fs");
const path_1 = require("path");
function detectYarnLockVersionByFile(file) {
    // @ts-ignore
    const fd = (0, fs_1.openSync)(file);
    const buf = Buffer.alloc(160);
    (0, fs_1.readSync)(fd, buf, 0, 160, 0);
    return (0, detectYarnLockVersion_1.detectYarnLockVersion)(buf);
}
function detectYarnLockVersionByDir(dir) {
    return detectYarnLockVersionByFile((0, path_1.join)(dir, 'yarn.lock'));
}
exports.default = detectYarnLockVersionByFile;
//# sourceMappingURL=detectYarnLockVersionByFile.js.map