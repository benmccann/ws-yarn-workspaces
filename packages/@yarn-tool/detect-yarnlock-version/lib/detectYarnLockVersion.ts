import { detectYarnLockVersionByObject } from './detectYarnLockVersionByObject';
import { EnumDetectYarnLock, EnumDetectYarnLockInputType } from '@yarn-tool/yarnlock-types';
import { parseYarnLockRawV2 } from '@yarn-tool/yarnlock-parse-raw';

export function _detectYarnLockVersionSimple(buf: Buffer | string)
{
	const head = buf.slice(0, 160).toString().trim();

	if (head.includes('# yarn lockfile v1'))
	{
		return EnumDetectYarnLock.v1
	}
	else if (/^__metadata:\s*version: (4|5)(?:\r|\n)/m.test(head))
	{
		return EnumDetectYarnLock.v2
	}
	else if (/^__metadata:\s*version: (\d)(?:\r|\n)/m.test(head))
	{
		return EnumDetectYarnLock.v3
	}

	return EnumDetectYarnLock.unknown
}

export function _detectYarnLockVersionCore<T extends Buffer | string>(input: T)
{
	let verType: EnumDetectYarnLock = _detectYarnLockVersionSimple(input);

	if (verType)
	{
		return {
			verType,
			detectType: EnumDetectYarnLockInputType.simple as const,
			input,
		}
	}

	if (verType = _tryParse(input))
	{
		return {
			verType,
			detectType: EnumDetectYarnLockInputType.parse_raw as const,
			input,
		}
	}

	return {
		verType: EnumDetectYarnLock.unknown as const,
		detectType: EnumDetectYarnLockInputType.unknown as const,
		input,
	}
}

export function detectYarnLockVersion(buf: Buffer | string)
{
	return _detectYarnLockVersionCore(buf).verType
}

/**
 * only check v2 and v3
 */
export function _tryParse(buf: Buffer | string)
{
	try
	{
		const yarnLockObject = parseYarnLockRawV2(buf.toString());
		return _tryParseObject(yarnLockObject)
	}
	catch (e)
	{

	}
}

export function _tryParseObject(yarnLockObject: Record<string, any>)
{
	const result = detectYarnLockVersionByObject(yarnLockObject);

	return (result === EnumDetectYarnLock.v2 || result === EnumDetectYarnLock.v3) ? result : void 0
}

export default detectYarnLockVersion;
