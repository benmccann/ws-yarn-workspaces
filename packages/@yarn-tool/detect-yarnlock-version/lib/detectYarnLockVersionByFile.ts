import { detectYarnLockVersion } from './detectYarnLockVersion';
import { openSync, readSync } from 'fs';
import { join } from 'path';

export function detectYarnLockVersionByFile(file: string)
{
	// @ts-ignore
	const fd = openSync(file)

	const buf = Buffer.alloc(160);

	readSync(fd, buf, 0, 160, 0)

	return detectYarnLockVersion(buf)
}

export function detectYarnLockVersionByDir(dir: string)
{
	return detectYarnLockVersionByFile(join(dir, 'yarn.lock'))
}

export default detectYarnLockVersionByFile
