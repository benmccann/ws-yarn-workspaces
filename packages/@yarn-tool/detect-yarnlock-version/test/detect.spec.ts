import { readFileSync } from "fs";
import { join, dirname, basename, extname } from "path";
import { detectYarnLockVersion, _detectYarnLockVersionCore, _tryParse } from '../lib/detectYarnLockVersion';
import { detectYarnLockVersionByFile, detectYarnLockVersionByDir } from '../lib/detectYarnLockVersionByFile';
import { __TEST_YARNLOCK } from '../../../../__root_ws';
import { sync as FastGlob } from '@bluelovers/fast-glob';
import { EnumDetectYarnLock } from '@yarn-tool/yarnlock-types';
import { _forEachVersionTags } from '../../../../test/lib/forEachVersionTags';

const __res = __TEST_YARNLOCK;

test(`empty`, () =>
{
	let file = join(__res, 'empty', 'yarn.lock')
	let buf = readFileSync(file)

	let actual = _detectYarnLockVersionCore(buf);
	let expected = EnumDetectYarnLock.unknown;

	expect(actual).toHaveProperty('verType', expected);

	expect(detectYarnLockVersionByFile(file)).toStrictEqual(expected);
	expect(detectYarnLockVersionByDir(dirname(file))).toStrictEqual(expected);

	expect(actual).toMatchSnapshot({
		input: expect.anything(),
	});

});

describe(`version`, () =>
{

	_forEachVersionTags().forEach(ver =>
	{
		describe(ver, () =>
		{
			const dir = join(__TEST_YARNLOCK, ver);

			FastGlob([
				'**/*.lock',
			], {
				cwd: dir,
			}).forEach(name => {
				const file = join(dir, name);
				const buf = readFileSync(file);

				test(name, () =>
				{

					let actual = _detectYarnLockVersionCore(buf);
					let expected = EnumDetectYarnLock[ver];

					expect(actual).toHaveProperty('verType', expected);

					expect(detectYarnLockVersionByFile(file)).toStrictEqual(expected);
					expect(detectYarnLockVersionByDir(dirname(file))).toStrictEqual(expected);

					if (ver !== 'v1')
					{
						expect(_tryParse(buf)).toStrictEqual(expected);
					}

					expect(actual).toMatchSnapshot({
						input: expect.anything(),
					});

				});
			})
		})

	});

})

describe(`others`, () =>
{
	FastGlob([
		'**/*.lock',
		'!v*/*',
	], {
		cwd: __TEST_YARNLOCK,
	}).forEach(name =>
	{
		const file = join(__TEST_YARNLOCK, name);
		const buf = readFileSync(file);

		test(name, () =>
		{
			let actual = _detectYarnLockVersionCore(buf);

			if (actual.verType && actual.verType !== EnumDetectYarnLock.v1)
			{
				expect(_tryParse(buf)).toStrictEqual(actual.verType);
			}

			expect(actual).toMatchSnapshot({
				input: expect.anything(),
			});

		});
	})
});
