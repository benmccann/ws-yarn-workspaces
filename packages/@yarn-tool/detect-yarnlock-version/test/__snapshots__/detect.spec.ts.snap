// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`empty 1`] = `
Object {
  "detectType": 0,
  "input": Anything,
  "verType": 0,
}
`;

exports[`others dedupe/empty/yarn.lock 1`] = `
Object {
  "detectType": 0,
  "input": Anything,
  "verType": 0,
}
`;

exports[`others dedupe/v1/next.js.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`others dedupe/v1/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`others dedupe/v2/456.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 2,
}
`;

exports[`others dedupe/v2/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 2,
}
`;

exports[`others diff/a/v1.iconv.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`others diff/a/v2.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 2,
}
`;

exports[`others diff/a/v3/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;

exports[`others diff/a/v3/yarn-2.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;

exports[`others diff/a/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`others diff/b/v1.iconv.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`others diff/b/v2.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 2,
}
`;

exports[`others diff/b/v3/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;

exports[`others diff/b/v3/yarn-2.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;

exports[`others diff/b/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`others empty.lock 1`] = `
Object {
  "detectType": 0,
  "input": Anything,
  "verType": 0,
}
`;

exports[`others empty/yarn.lock 1`] = `
Object {
  "detectType": 0,
  "input": Anything,
  "verType": 0,
}
`;

exports[`others ncu/v1/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`others ncu/v3/yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;

exports[`version v1 yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 1,
}
`;

exports[`version v2 meta-only.lock 1`] = `
Object {
  "detectType": 2,
  "input": Anything,
  "verType": 2,
}
`;

exports[`version v2 yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 2,
}
`;

exports[`version v3 meta-only.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;

exports[`version v3 yarn.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;

exports[`version v3 yarn-version-7.lock 1`] = `
Object {
  "detectType": 1,
  "input": Anything,
  "verType": 3,
}
`;
